articles/113_ros_time.md

```diff
@@ -0,0 +1,154 @@
+
+`SystemTime` will be directly tied to the system clock.
+
+### Steady Time
+
+Example use cases for this include hardware drivers which are interacting with peripherals with hardware timeouts.
+
+In nodes which require the use of `SteadyTime` or `SystemTime` for interacting with hardware or other peripherals it is expected that they do a best effort to isolate any `SystemTime` or `SteadyTime` information inside their implementation and translate external interfaces to use the ROS time abstraction when communicating over the ROS network.
+
+### ROS Time
+
+The `ROSTime` will report the same as `SystemTime` when an external Time Source is not active.
+When the external time source is active `ROSTime` will return the latest value reported by the Time Source.
+
+### Time Sources
+
+#### Default Time Source
+
+To implement the time abstraction the following approach will be used.
+
+The time abstraction can be published by one source on the `/clock` topic.
+The topic will contain the most up to date time for the ROS system.
+If a publisher exists for the topic, it will override the system time when using the ROS time abstraction.
+If `/clock` is being published, calls to the ROS time abstraction will return the latest time received from the `/clock` topic.
+If time has not been set it will return zero if nothing has been received.
+A time value of zero should be considered an error meaning that time is uninitialized.
+
+If the time on the clock jumps backwards, a callback handler will be invoked and be required to complete before any calls to the ROS time abstraction report the new time.
+Calls that come in before that must block.
+The developer has the opportunity to register callbacks with the handler to clear any state from their system if necessary before time will be in the past.
+
+The frequency of publishing the `/clock` as well as the granularity are not specified as they are application specific.
+
+##### No Advanced Estimating Clock
+
+There are more advanced techniques which could be included to attempt to estimate the propagation properties and extrapolate between time ticks.
+However all of these techniques will require making assumptions about the future behavior of the time abstraction.
+And in the case that playback or simulation is instantaneously paused, it will break any of these assumptions.
+There are techniques which would allow potential interpolation, however to make these possible it would require providing guarantees about the continuity of time into the future.
+For more accuracy the progress of time can be slowed, or the frequency of publishing can be increased.
+Tuning the parameters for the `/clock` topic lets you trade off time for computational effort and/or bandwidth.
+
+#### Custom Time Source
+
+It is possible that the user may have access to an out of band time source which can provide better performance than the default source the `/clock` topic.
+A potential alternative source of time might be a GPS.
```

Comment:
I'm not sure this is worth mentioning since the preferred way to integrate a GPS would be to integrate it into NTP. A better example would be to use a time synchronization algorithm (like ntp) with simulated time. That's a bit complicated to make clear here, so I'd suggest just leaving it at you can replace the time source for ROSTime with any arbitrary time source or algorithm.